File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/repository/MediathekRepository.java
package de.christinecoenen.code.zapp.app.mediathek.repository;

import androidx.paging.DataSource;

import org.joda.time.DateTime;

import java.util.Collections;
import java.util.List;

import de.christinecoenen.code.zapp.app.mediathek.api.MediathekService;
import de.christinecoenen.code.zapp.app.mediathek.api.request.QueryRequest;
import de.christinecoenen.code.zapp.app.mediathek.model.DownloadStatus;
import de.christinecoenen.code.zapp.app.mediathek.model.MediathekShow;
import de.christinecoenen.code.zapp.app.mediathek.model.PersistedMediathekShow;
import de.christinecoenen.code.zapp.persistence.Database;
import de.christinecoenen.code.zapp.utils.api.UserAgentInterceptor;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ConnectionSpec;
import okhttp3.OkHttpClient;
import okhttp3.TlsVersion;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;

public class MediathekRepository {


	private final MediathekService service;
	private final Database database;


	public MediathekRepository(Database database) {
		this.database = database;

		// workaround to avoid SSLHandshakeException on Android 7 devices
		// see: https://stackoverflow.com/questions/39133437/sslhandshakeexception-handshake-failed-on-android-n-7-0
		ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
			.tlsVersions(TlsVersion.TLS_1_1, TlsVersion.TLS_1_2, TlsVersion.TLS_1_3)
			.build();

		OkHttpClient client = new OkHttpClient.Builder()
			.connectionSpecs(Collections.singletonList(spec))
			.addInterceptor(new UserAgentInterceptor())
			.build();

		Retrofit retrofit = new Retrofit.Builder()
			.baseUrl("https://mediathekviewweb.de/api/")
			.client(client)
			.addConverterFactory(GsonConverterFactory.create())
			.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
			.build();

		service = retrofit.create(MediathekService.class);
	}

	public Single<List<MediathekShow>> listShows(@Body QueryRequest queryRequest) {
		return service.listShows(queryRequest)
			.subscribeOn(Schedulers.io())
			.map(mediathekAnswer -> {
				if (mediathekAnswer == null || mediathekAnswer.result == null || mediathekAnswer.err != null) {
					throw new RuntimeException("Empty result");
				}
				return mediathekAnswer.result.results;
			});
	}

	public DataSource.Factory<Integer, PersistedMediathekShow> getDownloads() {
		// TODO: filter for only queued, running or finished downloads
		// TODO: sort on date
		return database.mediathekShowDao().getAllDownloads();
	}

	public Flowable<PersistedMediathekShow> persistOrUpdateShow(MediathekShow show) {
		return Completable.fromAction(() -> database.mediathekShowDao().insertOrUpdate(show))
			.andThen(database.mediathekShowDao().getFromApiId(show.getApiId()))
			.subscribeOn(Schedulers.io());
	}

	public void updateShow(PersistedMediathekShow show) {
		database.mediathekShowDao()
			.update(show)
			.subscribeOn(Schedulers.io())
			.subscribe();
	}

	public void updateDownloadStatus(int downloadId, DownloadStatus downloadStatus) {
		database.mediathekShowDao()
			.updateDownloadStatus(downloadId, downloadStatus)
			.subscribeOn(Schedulers.io())
			.subscribe();
	}

	public void updateDownloadProgress(int downloadId, int progress) {
		database.mediathekShowDao()
			.updateDownloadProgress(downloadId, progress)
			.subscribeOn(Schedulers.io())
			.subscribe();
	}

	public void updateDownloadedVideoPath(int downloadId, String videoPath) {
		database.mediathekShowDao()
			.updateDownloadedVideoPath(downloadId, videoPath)
			.subscribeOn(Schedulers.io())
			.subscribe();
	}

	public Flowable<PersistedMediathekShow> getPersistedShow(int id) {
		return database.mediathekShowDao().getFromId(id).subscribeOn(Schedulers.io());
	}

	public Flowable<PersistedMediathekShow> getPersistedShowByApiId(String apiId) {
		return database.mediathekShowDao().getFromApiId(apiId).subscribeOn(Schedulers.io());
	}

	public Flowable<PersistedMediathekShow> getPersistedShowByDownloadId(int downloadId) {
		return database.mediathekShowDao().getFromDownloadId(downloadId).subscribeOn(Schedulers.io());
	}

	public Flowable<DownloadStatus> getDownloadStatus(String apiId) {
		return database
			.mediathekShowDao()
			.getDownloadStatus(apiId)
			.startWith(DownloadStatus.NONE)
			.subscribeOn(Schedulers.io());
	}

	public Flowable<Integer> getDownloadProgress(String apiId) {
		return database
			.mediathekShowDao()
			.getDownloadProgress(apiId)
			.subscribeOn(Schedulers.io());
	}

	public Single<Long> getPlaybackPosition(int showId) {
		return database.mediathekShowDao()
			.getPlaybackPosition(showId)
			.subscribeOn(Schedulers.io());
	}

	public void setPlaybackPosition(int showId, long positionMillis, long durationMillis) {
		database.mediathekShowDao()
			.setPlaybackPosition(showId, positionMillis, durationMillis, DateTime.now())
			.subscribeOn(Schedulers.io())
			.subscribe();
	}

	public Flowable<Float> getPlaybackPositionPercent(String apiId) {
		return database.mediathekShowDao()
			.getPlaybackPositionPercent(apiId)
			.startWith(0f)
			.distinct()
			.subscribeOn(Schedulers.io());
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/repositories/MediathekRepository.kt
package de.christinecoenen.code.zapp.repositories

import androidx.paging.PagingSource
import de.christinecoenen.code.zapp.models.shows.DownloadStatus
import de.christinecoenen.code.zapp.models.shows.MediathekShow
import de.christinecoenen.code.zapp.models.shows.PersistedMediathekShow
import de.christinecoenen.code.zapp.models.shows.SortableMediathekShow
import de.christinecoenen.code.zapp.persistence.Database
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.withContext
import org.joda.time.DateTime

class MediathekRepository(private val database: Database) {

	fun getDownloads(limit: Int): Flow<List<MediathekShow>> {
		return database
			.mediathekShowDao()
			.getDownloads(limit)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getDownloads(searchQuery: String): PagingSource<Int, SortableMediathekShow> {
		return database
			.mediathekShowDao()
			.getAllDownloads("%$searchQuery%")
	}

	fun getStarted(limit: Int): Flow<List<MediathekShow>> {
		return database
			.mediathekShowDao()
			.getStarted(limit)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getStarted(searchQuery: String): PagingSource<Int, SortableMediathekShow> {
		return database
			.mediathekShowDao()
			.getAllStarted("%$searchQuery%")
	}

	fun getBookmarked(limit: Int): Flow<List<MediathekShow>> {
		return database
			.mediathekShowDao()
			.getBookmarked(limit)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getBookmarked(searchQuery: String): PagingSource<Int, SortableMediathekShow> {
		return database
			.mediathekShowDao()
			.getAllBookarked("%$searchQuery%")
	}

	suspend fun persistOrUpdateShow(show: MediathekShow): Flow<PersistedMediathekShow> =
		withContext(Dispatchers.IO) {
			database
				.mediathekShowDao()
				.insertOrUpdate(show)

			database
				.mediathekShowDao()
				.getFromApiId(show.apiId)
				.distinctUntilChanged()
				.flowOn(Dispatchers.IO)
		}

	suspend fun updateShow(show: PersistedMediathekShow?) = withContext(Dispatchers.IO) {
		database
			.mediathekShowDao()
			.update(show!!)
	}

	suspend fun updateDownloadProgress(downloadId: Int, progress: Int) =
		withContext(Dispatchers.IO) {
			database
				.mediathekShowDao()
				.updateDownloadProgress(downloadId, progress)
		}

	suspend fun updateDownloadedVideoPath(downloadId: Int, videoPath: String?) =
		withContext(Dispatchers.IO) {
			database
				.mediathekShowDao()
				.updateDownloadedVideoPath(downloadId, videoPath)
		}

	fun getPersistedShow(id: Int): Flow<PersistedMediathekShow> {
		return database
			.mediathekShowDao()
			.getFromId(id)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getCompletedDownloads(): Flow<List<PersistedMediathekShow>> {
		return database
			.mediathekShowDao()
			.getCompletedDownloads()
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getPersistedShowByApiId(apiId: String): Flow<PersistedMediathekShow> {
		return database
			.mediathekShowDao()
			.getFromApiId(apiId)
			.distinctUntilChanged()
			.filterNotNull()
			.flowOn(Dispatchers.IO)
	}

	fun getPersistedShowByDownloadId(downloadId: Int): Flow<PersistedMediathekShow> {
		return database
			.mediathekShowDao()
			.getFromDownloadId(downloadId)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getDownloadStatus(id: Int): Flow<DownloadStatus> {
		return database
			.mediathekShowDao()
			.getDownloadStatus(id)
			.distinctUntilChanged()
			.onStart { emit(DownloadStatus.NONE) }
			.flowOn(Dispatchers.IO)
	}

	fun getDownloadStatus(apiId: String): Flow<DownloadStatus> {
		return database
			.mediathekShowDao()
			.getDownloadStatus(apiId)
			.filterNotNull()
			.distinctUntilChanged()
			.onStart { emit(DownloadStatus.NONE) }
			.flowOn(Dispatchers.IO)
	}

	fun getDownloadProgress(id: Int): Flow<Int> {
		return database
			.mediathekShowDao()
			.getDownloadProgress(id)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getDownloadProgress(apiId: String): Flow<Int> {
		return database
			.mediathekShowDao()
			.getDownloadProgress(apiId)
			.filterNotNull()
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	fun getIsBookmarked(apiId: String): Flow<Boolean> {
		return database
			.mediathekShowDao()
			.getIsBookmarked(apiId)
			.filterNotNull()
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}

	suspend fun setBookmarked(apiId: String, isBookmarked: Boolean) = withContext(Dispatchers.IO) {
		database
			.mediathekShowDao()
			.updateIsBookmarked(apiId, isBookmarked, DateTime.now())
	}

	fun getIsRelevantForUser(apiId: String): Flow<Boolean> {
		return database
			.mediathekShowDao()
			.getIsRelevantForUser(apiId)
			.distinctUntilChanged()
			.map { it == true }
			.onStart { emit(false) }
			.flowOn(Dispatchers.IO)
	}

	suspend fun getPlaybackPosition(showId: Int): Long = withContext(Dispatchers.IO) {
		database
			.mediathekShowDao()
			.getPlaybackPosition(showId)
	}

	suspend fun setPlaybackPosition(showId: Int, positionMillis: Long, durationMillis: Long) =
		withContext(Dispatchers.IO) {
			database
				.mediathekShowDao()
				.setPlaybackPosition(showId, positionMillis, durationMillis, DateTime.now())
		}

	suspend fun resetPlaybackPosition(apiId: String) = withContext(Dispatchers.IO) {
		database
			.mediathekShowDao()
			.resetPlaybackPosition(apiId)
	}

	fun getPlaybackPositionPercent(apiId: String): Flow<Float> {
		return database
			.mediathekShowDao()
			.getPlaybackPositionPercent(apiId)
			.distinctUntilChanged()
			.filterNotNull()
			.onStart { emit(0f) }
			.flowOn(Dispatchers.IO)
	}

	fun getCompletetlyDownloadedVideoPath(apiId: String): Flow<String?> {
		return database
			.mediathekShowDao()
			.getCompletetlyDownloadedVideoPath(apiId)
			.distinctUntilChanged()
			.flowOn(Dispatchers.IO)
	}
}
